sbom/spdx/encoder.go

```diff
@@ -0,0 +1,417 @@
+	v2common "github.com/spdx/tools-golang/spdx/v2/common"
+	"github.com/spdx/tools-golang/spdx/v2/v2_3"
+)
+
+type Version string
+
+const (
+	V2_3 Version = "v2.3"
+)
+
+type Format string
+
+const JSON Format = "json"
+
+// Creator describes the creator of the SPDX document that will be produced from the encoding.
+type Creator struct {
+	// Creator is the value of the [Creator] relationship.
+	Creator string
+	// CreatorType is the key of the [Creator] relationship.
+	// In accordance to the SPDX v2 spec, CreatorType should be one of "Person", "Organization", or "Tool".
+	CreatorType string
+}
+
+var _ sbom.Encoder = (*Encoder)(nil)
+
+type Encoder struct {
+	Version           Version
+	Format            Format
+	Creators          []Creator
+	DocumentName      string
+	DocumentNamespace string
+	DocumentComment   string
+}
+
+func NewDefaultEncoder(documentName, documentNamespace, documentComment string) *Encoder {
+	return &Encoder{
+		Version: V2_3,
+		Format:  JSON,
+		Creators: []Creator{
+			{
+				Creator:     fmt.Sprintf("Claircore-%s", getVersion()),
+				CreatorType: "Tool",
+			},
+		},
+		DocumentName:      documentName,
+		DocumentNamespace: documentNamespace,
+		DocumentComment:   documentComment,
+	}
+}
+
+// Encode encodes a [claircore.IndexReport] to an [io.Reader].
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, w io.Writer, ir *claircore.IndexReport) error {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return err
+	}
+
+	// TODO(blugo): support SPDX versions before 2.3
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return fmt.Errorf("unknown SPDX version: %v", e.Version)
+	}
+
+	switch e.Format {
+	case JSON:
+		if err := spdxjson.Write(tmpConverterDoc, w); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return fmt.Errorf("unknown requested format: %v", e.Format)
+}
+
+func (e *Encoder) parseIndexReport(ctx context.Context, ir *claircore.IndexReport) (*v2_3.Document, error) {
+	creators := make([]v2common.Creator, len(e.Creators))
+	for i, creator := range e.Creators {
+		creators[i] = v2common.Creator{
+			Creator:     creator.Creator,
+			CreatorType: creator.CreatorType,
+		}
+	}
+
+	// Initial metadata
+	out := &v2_3.Document{
+		SPDXVersion:       v2_3.Version,
+		DataLicense:       v2_3.DataLicense,
+		SPDXIdentifier:    "DOCUMENT",
+		DocumentName:      e.DocumentName,
+		DocumentNamespace: e.DocumentNamespace,
+		CreationInfo: &v2_3.CreationInfo{
+			Creators: creators,
+			Created:  time.Now().Format("2006-01-02T15:04:05Z"),
+		},
+		DocumentComment: e.DocumentComment,
+	}
+
+	pkgMap := map[int]*v2_3.Package{}
```

Comment:
won't block the PR but I tend to prefer `pkgMap := make(map[int]*v2_3.Package)` though I do see this pattern throughout the Claircore codebase

Also maybe rename to `pkgs`?